A59-audit-logging.md

```diff
@@ -0,0 +1,243 @@
+A59: gRPC Audit Logging
+----
+* Author(s): [Luwei Ge](https://github.com/rockspore)
+* Approver:
+* Status: Draft {Draft, In Review, Ready for Implementation, Implemented}
+* Implemented in:
+* Last updated: 2023-03-09
+* Discussion at: 
+
+## Abstract
+
+Support audit logging based on authorization events in gRPC servers in both
+xDS-enabled and gRPC authorization policy based use cases.
+
+## Background
+
+gRPC's authorization support is based on the [RBAC policy][]. In
+xDS-enabled cases, users can configure authorization policies in the control
+plane which applies [RBAC HTTP filters][RBAC filter] to the gRPC servers. When
+xDS is not an option, users can use the [gRPC authorization policy][A43] to
+get similar authorization enforment.
+
+This proposal describes the audit logging APIs that allow users to audit
+specific authorization events in their gRPC servers with the necessary context.
+We aim to offer consistent APIs to support both aforementioned use cases.
+
+[RBAC filter]: https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/http/rbac/v3/rbac.proto
+[RBAC policy]: https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/rbac/v3/rbac.proto
+[HttpConnectionManager proto]: https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
+
+### Related Proposals: 
+* [gRFC A41: xDS RBAC Support][A41]
+* [gRFC A43: gRPC Authorization API][A43]
+
+[A41]: A41-xds-rbac.md
+[A43]: A43-grpc-authorization-api.md
+
+## Proposal
+
+### Audit Context
+
+The table below lists the metadata we will include in the audit context,
+which will be made available to audit loggers during an audit event.
+
+It is worth noting that audit happens right after the authorization
+decision is made, so the status only reflects that but not the eventual
+one from the server. 
+
+|Fields             |Additional Information                                                             |Example                |
+|-------------------|-----------------------------------------------------------------------------------|-----------------------|
+|RPC Method         |Method the RPC hit.                                                                |"/pkg.service/foo"     |
+|Principal          |Identity the RPC. Currently only available in certificate-based TLS authentication.|"spiffe://foo/user1"   |
+|Timestamp          |Time when the RPC happens.                                                         |1649376211             |
```

Comment:
What does "happens" mean here?  The time at which the server first saw the RPC?  Or the time at which the authz policy made the decision about it?  Those two things will not happen at the same moment.

If it's the time that the authz policy made the decision, then do we actually need to make this available in the audit context?  It seems like the audit logger impl could just check the current time itself.

Reply:
I revised the description to be `when the server saw the RPC`. For audit logging, my understanding is we generally want to know when the RPC was attempted not when the authorization decision is made or to be made.

Reply:
Is it actually important to distinguish between the moment that the server first saw the RPC and the moment that the authz decision is made?  Note that the authz decision is made as soon as the server sees the initial metadata from the client, which is always the first operation on the call.  So there are only a small set of cases in which these two timestamps will be significantly different (e.g., if for some reason the transport reads the start of the initial HEADERS frame but then there's a long delay before it reads the rest of the HEADERS frame), and I'm not sure anyone will care about that from the perspective of the audit logs.

My inclination here is to just remove this field and let the logger use the current time for the log entry.

If we do include this field here, then I think we need to be very specific about what timestamp it indicates, because "when the server saw the RPC" is not very precise.  I think it would need to say something like "the time at which the transport reads the start of the stream" or "the time at which the call is created".  And I think the need for being this precise is actually a good reason not to include this here, since it will ultimately over-specify the design, which may unnecessarily constrain implementation choices.

Reply:
Thanks, Mark, for the detailed explanation. As we also discussed offline, we should be fine by removing this timestamp field and have the logger to use its current time. Change has been made.